[FLINK-24987][Configuration] Add explicit enum value NO_EXTERNAL_CHECKPOINTS as default for ExternalizedCheckpointRetention

nicoweidner opened the pull request:
## What is the purpose of the change

*Add NO_EXTERNAL_CHECKPOINTS as an explicit enum value and set it as the default for ExternalizedCheckpointCleanup. This clarifies its current usage and allows users to change from either of the other two options to NO_EXTERNAL_CHECKPOINTS without having to deal with null values.*

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.
Current behavior is unchanged.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? Generated docs reflect the new default value

Airblader commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/CheckpointConfig.java:
```diff
@@ -446,7 +447,7 @@ public void setTolerableCheckpointFailureNumber(int tolerableCheckpointFailureNu
      */
     @PublicEvolving
-    public void enableExternalizedCheckpoints(ExternalizedCheckpointCleanup cleanupMode) {
+    public void setExternalizedCheckpointCleanup(ExternalizedCheckpointCleanup cleanupMode) {
```
Renaming this is a breaking change. We should probably keep the old method, too, and deprecate it.

zentol replied:
In fact it broke the python api :)

`
Nov 24 11:28:23 FAILED pyflink/datastream/tests/test_check_point_config.py::CheckpointConfigTests::test_get_set_externalized_checkpoints_cleanup`

nicoweidner replied:
Woops. Thanks, that's something I am obviously not yet sensitive to. I will keep the old one and let it call a new method.

On the note of `@Public` and `@PublicEvolving`: CheckpointConfig has `@Public`, but some of its methods have `@PublicEvolving`. How does that make sense? On behalf of the former, I would expect that all methods of the class must have stable interfaces.
This might be related to there not being a clear opinion on whether these annotations are "inherited" by the methods.

Airblader replied:
It is allowed for individual methods to have a different stability. Basically, methods inherit the stability of their class unless they are annotated themselves. Otherwise extending public APIs with new methods would be very difficult as you could only do it by immediately having new methods be stable.

nicoweidner replied:
The Python API issues should be resolved now, there were in fact a few more problems than just this renaming.

Airblader commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/CheckpointConfig.java:
```diff
@@ -711,7 +713,7 @@ public long getCheckpointIdOfIgnoredInFlightData() {
          * JobStatus#FAILED}.
          */
-        DELETE_ON_CANCELLATION(true),
+        DELETE_ON_CANCELLATION,
```
Since we're at it, we could have the enum implement `DescribedEnum` and clean up the description of the config option to not mention the enum values in the text (this will then be generated from the enum).

Airblader commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/ExecutionCheckpointingOptions.java:
```diff
@@ -88,16 +88,18 @@
 
     public static final ConfigOption<CheckpointConfig.ExternalizedCheckpointCleanup>
-            EXTERNALIZED_CHECKPOINT =
+            EXTERNALIZED_CHECKPOINT_CLEANUP =
```
This is a breaking change. If it isn't necessary, we should avoid it; otherwise we can keep a deprecated alias for now.

Airblader commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/ExecutionCheckpointingOptions.java:
```diff
@@ -88,16 +88,18 @@
                                                     "Externalized checkpoints write their meta data out to persistent storage and are not "
                                                             + "automatically cleaned up when the owning job fails or is suspended (terminating with job "
-                                                            + "status %s or %s. In this case, you have to manually clean up the checkpoint state, both the "
+                                                            + "status %s or %s). In this case, you have to manually clean up the checkpoint state, both the "
```
As mentioned above, we should not describe the options here but let the enum do that.

nicoweidner replied:
I am implementing `DescribedEnum` now to add some more description, but just for the record: The enum values are not mentioned here - `JobStatus` is a different enum...

nicoweidner replied:
I left this description as it does not really describe the options, but rather the consequences for the users (you have to clean up leftover checkpoints yourself). Do you think it's too much overlap?

zentol commented:
@pnowojski Do you see a problem with having an explicit value for disabling externalized checkpoints?

Airblader commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/CheckpointConfig.java:
```diff
@@ -437,15 +443,44 @@ public void setTolerableCheckpointFailureNumber(int tolerableCheckpointFailureNu
+
+    /**
+     * Sets the mode for externalized checkpoint clean-up. Externalized checkpoints will be enabled
+     * * automatically unless the mode is set to {@link
```
```suggestion
     * automatically unless the mode is set to {@link
```

Airblader reviewed (approved):
LGTM overall, just one nit for the docs still

nicoweidner commented:
@flinkbot run azure

pnowojski commented:
> @pnowojski Do you see a problem with having an explicit value for disabling externalized checkpoints?

I don't see potential problems with this, as long as we maintain backward compatibility (which I think we do in this case). And the proposal overall makes sense to me.

Airblader reviewed (changes requested):
Thanks @nicoweidner. I can take care of merging this PR. Can you rebase the PR to resolve the conflict and add the (primary) components to each commit, i.e. streaming-java and docs, correspondingly? Thanks!

nicoweidner commented:
> Thanks @nicoweidner. I can take care of merging this PR. Can you rebase the PR to resolve the conflict and add the (primary) components to each commit, i.e. streaming-java and docs, correspondingly? Thanks!

Thanks @Airblader! I rebased and added the components as you suggested

Airblader commented:
@flinkbot run azure

Airblader reviewed (approved):
LGTM
